compiler/src/main/kotlin/io/spine/protodata/codegen/BaseTypeSystem.kt

```diff
@@ -0,0 +1,172 @@
+import io.spine.protodata.Value.KindCase.DOUBLE_VALUE
+import io.spine.protodata.Value.KindCase.ENUM_VALUE
+import io.spine.protodata.Value.KindCase.INT_VALUE
+import io.spine.protodata.Value.KindCase.LIST_VALUE
+import io.spine.protodata.Value.KindCase.MAP_VALUE
+import io.spine.protodata.Value.KindCase.MESSAGE_VALUE
+import io.spine.protodata.Value.KindCase.NULL_VALUE
+import io.spine.protodata.Value.KindCase.STRING_VALUE
+
+/**
+ * The type system of an application being built by ProtoData.
+ *
+ * The type system knows the association between the Protobuf types and the types of
+ * a target language.
+ *
+ * @param T the type of a language-specific type name
+ * @param V the type of a language-specific expression, i.e. the code that yields a value
+ */
+public abstract class BaseTypeSystem<T: CodePrintable, V: CodePrintable>
+protected constructor(
+    private val knownTypes: Map<TypeName, T>
+) {
+
+    /**
+     * A [ValueConverter] for converting Protobuf values into language-specific code expressions.
+     */
+    protected abstract val valueConverter: ValueConverter<V>
+
+    /**
+     * Converts the given Protobuf type into a language-specific type name.
+     */
+    public fun typeNameToCode(type: Type): T {
+        return when {
+            type.hasPrimitive() -> convertPrimitiveType(type.primitive)
+            type.hasMessage() -> convertTypeName(type.message)
+            type.hasEnumeration() -> convertTypeName(type.enumeration)
+            else -> unknownType(type)
+        }
+    }
+
+    /**
+     * Converts the given Protobuf primitive type into a language-specific type name.
+     *
+     * For example, `PrimitiveType.TYPE_INT64` is converted into `long` in Java.
+     */
+    protected abstract fun convertPrimitiveType(type: PrimitiveType): T
+
+    /**
+     * Converts the given Protobuf type name into a language-specific type name.
+     *
+     * The type name can represent a message or an enum type.
+     */
+    public fun convertTypeName(protoName: TypeName): T =
+        knownTypes[protoName] ?: unknownType(protoName)
+
+    /**
+     * Converts the given Protobuf value into a language-specific code expression.
+     *
+     * @see ValueConverter
+     */
+    public fun valueToCode(value: Value): V = with(valueConverter) {
+        when (value.kindCase) {
+            NULL_VALUE -> toNull(value)
+            BOOL_VALUE -> toBool(value)
+            DOUBLE_VALUE -> toDouble(value)
+            INT_VALUE -> toInt(value)
+            STRING_VALUE -> toString(value)
+            BYTES_VALUE -> toBytes(value)
+            MESSAGE_VALUE -> toMessage(value)
+            ENUM_VALUE -> toEnum(value)
+            LIST_VALUE -> toList(value)
+            MAP_VALUE -> toMap(value)
+            else -> throw IllegalArgumentException("Empty value")
+        }
+    }
+
+    /**
+     * A factory of language-specific code, that represents a Protobuf value.
+     */
+    public interface ValueConverter<V: CodePrintable> {
```

Comment:
Why do we need this interface nested? We have `Value` as a type. The converter to code in a certain language comes somewhere near. It does not relate directory to a type system. It relates to a programming language and expressions in it.

Reply:
A `ValueConverter` is heavily intertwined with the `TypeSystem`. It needs the `TypeSystem` to know how to print a type's name in a language-specific manner.

The `TypeSystem` also serves as a public facade for the `ValueConverter`'s operations, so that the user has to deal with only one concept when developing code generation, not with two.

Reply:
I understand this is the consequence of the currently used decomposition. I asked the question because it looks as a design issue for me. Anyway, let's discuss it on Thursday.